core/src/main/java/com/graphhopper/routing/ch/PrepareContractionHierarchies.java

```diff
@@ -79,7 +79,9 @@ public static PrepareContractionHierarchies fromGraphHopperStorage(GraphHopperSt
     }
 
     private PrepareContractionHierarchies(GraphHopperStorage ghStorage, CHConfig chConfig) {
-        this.chGraph = ghStorage.getCHGraph(chConfig);
+        chGraph = ghStorage.getCHGraph(chConfig.getName());
+        if (chGraph == null)
+            throw new IllegalArgumentException("There is no CH graph '" + chConfig.getName() + "', existing: " + ghStorage.getCHGraphNames());
```

Comment:
Can we throw this inside ghStorage.getCHGraph?

Reply:
Good question :) Previously there was an `IllegalStateException` inside `getCHGraph()`, but what about the case where we get *some* profile via the web UI and use this to get the CH graph? In this case we do not want a 500 error, so we would have to catch the `IllegalStateException` and convert it into a (web api friendly(ier)) `IllegalArgumentException`, but this very ugly. So we could change this into an `IllegalArgumentException` instead, but this only means that the concerns of the web layer further propagate into the core (especially if the error message contains web layer stuff). I thought returning `null` is a cleaner (layer independent) way to communicate that the corresponding `CHGraph` does not exist. I would even use a (checked) `CHGraphDoesNotExistException` to make this very explicit, but I know you don't like it and we had this discussion before :)

To summarize:

- The `IllegalStateException` used before seems very wrong to me, because it does not even allow asking `GraphHopperStorage` for a wrong `CHGraph` name.
- Using `IllegalArgumentException` is slightly better, especially because we use this quite frequently so far. But what about the error message? Its hard to make this independent from the layer the method is used in. And catching and converting an `IllegalArgumentException` also seems wrong. In this case it would be better to use a dedicated exception.
- Using `null` leaves it up to the calling code how to handle the exceptional case (this is what I decided to do here)

Reply:
Ah. Thanks for these insights. So we fetch getCHGraph to know if a ch graph exists? Couldn't we do this with the profile map or when is it required?

> but I know you don't like it and we had this discussion before :)

Haha. Yeah. But every rule can have its exceptions ;)

Reply:
> So we fetch getCHGraph to know if a ch graph exists?

No, we need the actual graph (here and in other places). 

> Haha. Yeah. But every rule can have its exceptions ;)

No I think there is no need for this. Its a matter of style and since we aren't using checked exceptions there is no real need to start doing it. Sorry mentioning checked/unchecked was a bit distracting here. This is more about error handling between different parts of the code base. And I guess the question was whether we should use return values like `null`, or standard exceptions (like `IllegalArgument`) or dedicated exceptions (like `CHGraphDoesNotExist`) to indicate such exceptional conditions.

Reply:
> Using IllegalArgumentException is slightly better, especially because we use this quite frequently so far. But what about the error message? Its hard to make this independent from the layer the method is used in.

Are there currently different exception handling cases for `ghStorage.getCHGraph==null` where we have different messages?

(Because if there is a bug then the now unchecked cases like in createRouter will throw a NPE)

Reply:
> Are there currently different exception handling cases for ghStorage.getCHGraph==null where we have different messages?

We don't know what `GraphHopperStorage` might be used for so we cannot know what a good error message would be. Actually I think we do not want to know this because `GraphHopperStorage` is not meant to be used for one particular use-case. This is kind of my point here, it should be up to the caller of the method what to do if the CH graph does not exist. In case the CH graph does not exist and we use the result we get an NPE yes, but before we got an `IllegalStateException` (I do not think either one is better). Using `null` forces the caller to handle it unless she is sure this case cannot happen (compared to a checked exceptions where we also have to handle it (and the compiler tells us) if we are sure we do not need to. Ok, the NPE will not produce a useful error message, so this would be an advantage of using a dedicated exception.

Reply:
> like in createRouter

In createRouter there cannot be an NPE, because we use a profile name that we got from GraphHopperStorage (but yes this is hard to see and not so easy to reason about).

Reply:
Ah but maybe we are talking about two different things... Have a look at `Router#getAlgorithmFactory`

```java
   CHGraph chGraph = chGraphs.get(profile);
            if (chGraph == null)
                throw new IllegalArgumentException("Cannot find CH preparation for the requested profile: '" + profile + "'" +
                        "\nYou can try disabling CH using " + Parameters.CH.DISABLE + "=true" +
                        "\navailable CH profiles: " + chGraphs.keySet());
```

Here we get a profile name from the user and throw an error if this profile name does not work with CH. The same error message would not make sense in `GraphHopperStorage#getCHGraph`. But to make this even more complicated: This code no longer even takes the CH graph from `GraphHopperStorage` (but it did before), because now I have this `Map<String, CHGraph>` (see the comment in the `Router` constructor if you wonder why).

Reply:
But I can see why you want to prevent NPE, so my next suggestion would be an (unchecked) `CHGraphDoesNotExistException`.

Reply:
> So I could also revert the change in getCHGraph and throw an exception there again

ok

> But I can see why you want to prevent NPE, so my next suggestion would be an (unchecked) CHGraphDoesNotExistException

Let's keep it like it is :) (returning null)